File: backend/app/api/v1/kb.py
"""Knowledge-base CRUD endpoints."""
from fastapi import APIRouter, HTTPException, status

from app.core.deps import CurrentUser, DbSession
from app.schemas.kb import KBCreate, KBListResponse, KBResponse, KBUpdate
from app.services import kb_service

router = APIRouter(prefix="/knowledge-base", tags=["knowledge-base"])


@router.get("", response_model=KBListResponse)
async def list_kb(
    db: DbSession,
    user: CurrentUser,
    search: str | None = None,
    category: str | None = None,
    tag: str | None = None,
    limit: int = 50,
    offset: int = 0,
):
    rows, total = await kb_service.list_items(
        db,
        search=search,
        category=category,
        tag=tag,
        limit=min(limit, 200),
        offset=max(offset, 0),
    )
    return KBListResponse(items=[KBResponse.model_validate(r) for r in rows], total=total)


@router.get("/categories", response_model=list[str])
async def list_categories(db: DbSession, user: CurrentUser):
    return await kb_service.categories(db)


@router.post("", response_model=KBResponse, status_code=status.HTTP_201_CREATED)
async def create_kb(req: KBCreate, db: DbSession, user: CurrentUser):
    return await kb_service.create(db, user=user, req=req)


@router.get("/{item_id}", response_model=KBResponse)
async def get_kb(item_id: int, db: DbSession, user: CurrentUser):
    item = await kb_service.get(db, item_id=item_id)
    if not item:
        raise HTTPException(status_code=404, detail="Item not found.")
    return item


@router.patch("/{item_id}", response_model=KBResponse)
async def update_kb(item_id: int, req: KBUpdate, db: DbSession, user: CurrentUser):
    item = await kb_service.get(db, item_id=item_id)
    if not item:
        raise HTTPException(status_code=404, detail="Item not found.")
    return await kb_service.update(db, item=item, req=req)


@router.delete("/{item_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_kb(item_id: int, db: DbSession, user: CurrentUser):
    item = await kb_service.get(db, item_id=item_id)
    if not item:
        raise HTTPException(status_code=404, detail="Item not found.")
    await kb_service.delete(db, item=item)

File: frontend/src/api/kb.ts
import { api } from "./client";

export interface KBItem {
  id: number;
  owner_user_id: number | null;
  title: string;
  category: string;
  body: string;
  tags: string[];
  created_at: string;
  updated_at: string;
}

export interface KBListResponse {
  items: KBItem[];
  total: number;
}

export interface KBCreateInput {
  title: string;
  category: string;
  body: string;
  tags: string[];
}

export type KBUpdateInput = Partial<KBCreateInput>;

export interface KBListParams {
  search?: string;
  category?: string;
  tag?: string;
  limit?: number;
  offset?: number;
}

export async function listKB(params: KBListParams = {}): Promise<KBListResponse> {
  const res = await api.get<KBListResponse>("/knowledge-base", { params });
  return res.data;
}

export async function listCategories(): Promise<string[]> {
  const res = await api.get<string[]>("/knowledge-base/categories");
  return res.data;
}

export async function getKB(id: number): Promise<KBItem> {
  const res = await api.get<KBItem>(`/knowledge-base/${id}`);
  return res.data;
}

export async function createKB(input: KBCreateInput): Promise<KBItem> {
  const res = await api.post<KBItem>("/knowledge-base", input);
  return res.data;
}

export async function updateKB(id: number, input: KBUpdateInput): Promise<KBItem> {
  const res = await api.patch<KBItem>(`/knowledge-base/${id}`, input);
  return res.data;
}

export async function deleteKB(id: number): Promise<void> {
  await api.delete(`/knowledge-base/${id}`);
}
